Flag `default` switch statement condition clause as keyword

msftrncs opened the pull request:
# PR Summary

Flag the `switch` statement condition clause `default` token as a keyword
token to facilitate (PSReadLine) highlighting it as a keyword.  Closes #10470.

This change is to treat the `default` clause keyword as a full keyword token, and as such token based highlighters will see it as a keyword.

## PR Context

Flagging the `default` condition clause token of the `switch` statement as a keyword will facilitate highlighting of the keyword to better contrast against other bareword string literal conditions.

A simple (partial) example:
```PowerShell
# return simple element value (need to check for Boolean datatype, and process value accordingly)
switch ($node.Name) {
    true { $true; continue } # return a Boolean TRUE value
    false { $false; continue } # return a Boolean FALSE value
    default { $node.Value } # return the element value
}
```
Highlighting of the `default` keyword in the above example makes it easier to distinguish the special condition clause from the other bareword conditions.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

msftrncs commented:
I may need guidance on tests, I have not yet ventured in to any of the tests so far.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -6126,6 +6126,10 @@ private ExpressionAst GetCommandArgument(CommandArgumentContext context, Token t
                                 break;
+                            case CommandArgumentContext.SwitchCondition when token.Kind == TokenKind.Identifier && token.Text.Equals("default", StringComparison.OrdinalIgnoreCase):
+                                // `Default` is a keyword of very limited scope so do color it that way in this case
+                                token.TokenFlags |= TokenFlags.Keyword;
```
I believe we need to enhance `enum TokenKind`, `s_staticTokenFlags`, `s_keywordTokenKind`.

msftrncs replied:
@iSazonov, are you suggesting I try to implement `default` as a fully specified token?

- [x] Add to `enum.TokenKind`, `s_staticTokenFlags`, `s_tokenText`.
- [x] Add to `s_keywordText`, `s_keywordTokenKind`.
- [x] Correct assertions for table length checks.
- [ ] Use token in `switch` statement condition parsing(`parser`) logic.
- [ ] Implement tests …

It does not appear that I have to add any further logic to tokenizer, correct?

This way has higher difficulty than it might first appear, because now logic that was purposely put to prevent keywords from being detected in this position now has to be bypassed.  Or in other words, the code changes I already committed (seen here) are still needed to some extent, even after adding the `default` keyword to the keyword token list.

iSazonov replied:
> to implement default as a fully specified token

I think it was a conclusion in the issue.
@daxian-dbw @rjmholt Could you please share your thoughts before @msftrncs continue?

msftrncs replied:
I think I was able to modify the switch condition clause logic to utilize the keyword token, thus preventing the token from being reverted to a command argument, eliminating the need for extra logic in `GetCommandArgument` to prevent that.  However, not sure which way is more efficient.

msftrncs replied:
I do have a commit ready to post, implementing `default` as a full token, with revisions to SwitchStatementRule to specifically utilize it instead of a string comparison.  In that case, the change here is no longer needed and such would not be present.

iSazonov commented:
I do not find any tests for keywords.

msftrncs commented:
As for tests, it looks like, adding to `test\powershell\language\parser\parsing.tests.ps1`, some `ParseInput` expressions could be constructed to test the tokenizer results of a sample `switch` statement.  I am referencing the ternary operator parsing tests that @daxian-dbw added.

msftrncs committed:
Correct test for change of Default keyword.

Correct 'switch statement parsing' test for change in error output due to
`default` now being a keyword.

msftrncs commented:
I rebased, and have updated the commits to reflect creating the `default` keyword token.  I have also included a potential test, after correcting an existing one that changed because of the error records no longer being the same.  Please provide an initial review of this approach, and if there is any other testing that should be done.

iSazonov commented on src/System.Management.Automation/engine/parser/token.cs:
```diff
@@ -1125,16 +1129,17 @@ public static class TokenTraits
             /*              Command */ "command",
             /*               Hidden */ "hidden",
             /*                 Base */ "base",
+            /*                 Base */ "default",
```
```suggestion
            /*              Default */ "default",
```

msftrncs replied:
Opps

iSazonov commented on test/powershell/Language/Parser/Parsing.Tests.ps1:
```diff
@@ -423,3 +423,35 @@ Describe "Ternary Operator parsing" -Tags CI {
+    It "'default' can be used as command name" {
+        function default { 'default' }
+
+        default | Should -BeExactly 'default'
```
This test seems to be redundant since we already have tests for this case.

msftrncs replied:
I cannot find anywhere that this is specifically tested, and I am not really looking in the C# tests, so I may be missing it, either way, I do not know if there is value in this test or not.

iSazonov replied:
Perhaps @vexx32 (who pulled #9812) could correct me.

vexx32 replied:
I don't think this is quite the same case here. Those tests ensure that we can name a class method after a language keyword, which is a different context than this case — ensuring that a keyword is only valid in a constrained context, and is permitted as a command name outside that context.

I think this test is worthwhile. We don't have many constrained keywords (outside of the aforementioned exception where most keywords are permitted as class method names regardless of their usual restrictions).

iSazonov replied:
@vexx32 Many thanks! I was sure that we already have these tests.

@msftrncs Could you please enhance your test with Loel's suggesion? (TestCases helps.)

msftrncs replied:
I will reorganize the tests that I was adding here to be a more generic set of tests for 'limited scope' keywords.  I will include `in` as well, as its also a limited scope keyword, if I do not find any tests for it.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -2744,28 +2744,42 @@ private StatementAst SwitchStatementRule(LabelToken labelToken, Token switchToke
+                            // ErrorRecovery: if we don't have anything that looks like a condition, we won't
+                            // find a body (because a body is just a script block, which works as a condition.)
+                            // So don't look for a body, hope we find the '}' next.
+
```
Extra line.
```suggestion
```

msftrncs replied:
I did not add this extra line, I only indented the code block within another `if` block.   Do you still want the line removed?

iSazonov replied:
CodeFactor will complains on your commit. So please remove.

iSazonov commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -2744,28 +2744,42 @@ private StatementAst SwitchStatementRule(LabelToken labelToken, Token switchToke
+                            ReportIncompleteInput(After(endErrorStatement),
+                                nameof(ParserStrings.MissingSwitchConditionExpression),
+                                ParserStrings.MissingSwitchConditionExpression);
+                            // Consume a closing curly, if there is one, to avoid an extra error
```
Please add new line.
```suggestion

                            // Consume a closing curly, if there is one, to avoid an extra error
```

msftrncs replied:
Ditto, I only indented the code block within another `if` block.   Do you still want an blank line added?

iSazonov replied:
The same.

vexx32 replied:
@msftrncs for reference, if we update a section of code (even if just by indentation) CodeFactor will consider it part of the modifications of a PR and will raise a warning about comment styles. I believe the current comment style rule is that comments should have a blank line above, but not below them.

msftrncs replied:
@vexx32, I just wanted to be sure, sometimes the preference is not to modify blocks of code that are not actually being changed in any functional way, including in reformatting, other than changing the indent.   Different repositories have different tolerances to changes such as this.

iSazonov replied:
You can ignore such minor style issues. I will fix them later per month.

msftrncs committed:
Add `Default` TokenKind

Add `Default` token for switch statement `default` condition clause keyword,
allowing token based syntax highlighting to indicate `default` is a keyword.

Adjusted SwitchStatementRule to utilize new token.
clauseCondition is no longer produced for the `Default` clause.

msftrncs committed:
Add tests for parsing 'limited' keywords.

Add tests for parsing 'limited' keywords, `default`, `hidden`, `in`,
and `static` to check the token results as expected for syntax high-
lighting applications.  Validates the keywords can be used as function
names.

This is preliminary (WIP) concept.

msftrncs commented:
I have revised the commits again, though the bulk of changes is now in the tests.  I still have the feeling the tests are still a work in progress, though they are closer to what I was first expecting.

msftrncs commented:
I'm removing the WIP prefix, as I think this is ready for review.  I am quite interested for feedback.

iSazonov commented:
@daxian-dbw @JamesWTruher @adityapatwardhan Please review the PR.

rjmholt reviewed (commented):
Looks like a started a review on this PR some time ago but didn't post it... Comments left unchanged

rjmholt reviewed (approved):
Ok, the code here looks good to me. Thanks for adding thorough tests!

daxian-dbw committed:
Keep backward compatibility

daxian-dbw reviewed (approved):
@msftrncs Sorry that this PR has been overlooked for sooo long ...
The changes look good to me. I made some minor changes to keep the backward compatibility around the `ErrorStatement` case.

iSazonov commented:
@msftrncs Thanks for your contribution!
